src/engine/rand.h

```diff
@@ -29,12 +29,39 @@
 #include <cstdlib>
 #include <functional>
 #include <random>
+#include <string>
 #include <type_traits>
 #include <utility>
 #include <vector>
 
 namespace Rand
 {
+    uint32_t uniformIntDistribution( const uint32_t from, const uint32_t to, std::mt19937 & gen );
+
+    // Fisher-Yates shuffle AKA Knuth shuffle, probably the same as std::shuffle.
+    // NOTE: we can't use std::shuffle here because it uses std::uniform_int_distribution which behaves differently on different platforms.
+    template <class Iter, class Random>
+    void shuffle( Iter first, Iter last, Random && gen ) // NOLINT(cppcoreguidelines-missing-std-forward): same signature as std::shuffle
+    {
+        if ( first == last ) {
+            return;
+        }
+
+        assert( first < last );
+
+        // change last from once-past-the-end to last element
+        --last;
+        uint32_t remainingSwaps = static_cast<uint32_t>( last - first );
+        while ( remainingSwaps > 0 ) {
+            // Allow argument-dependant lookup (ADL) for swap: first try in the namespace of the type, then in the std namespace.
+            using std::swap;
+            const uint32_t index = uniformIntDistribution( 0, remainingSwaps, gen );
+            swap( *last, *( first + index ) );
+            --last;
+            --remainingSwaps;
+        }
```

Comment:
How about something like this:

```suggestion
        // last points to the past-the-end element that should not participate in the shuffling process
        for ( auto i = last - first - 1; i > 0; --i ) {
            using std::swap;
            swap( *( first + i ), *( first + uniformIntDistribution( 0, static_cast<uint32_t>( i ), gen ) ) );
        }
```

The same idea, but looks a bit shorter.

Reply:
As for me, this variant has the following objective advantages:

1. There is no need to handle the special case when `last == first`;
2. There is no need to additionally check that `to` is not less than `from`;
3. `first` and `last` can be `const`.

And there is also a subjective advantage that this variant is much shorter. As for clarity, verbosity is not always equivalent to clarity. However, if you prefer your variant, then I do not insist.

Reply:
>There is no need to handle the special case when last == first;
There is no need to additionally check that last is not less than first;

All true, but come at the cost of using int64 arithmetic rather than uint32_t. Probably doesn't matter on 64-bit CPU though ¯\_(ツ)_/¯

> Both first and last can be const
I actually like const 😄 . Didn't seem to matter much as arguments are passed by value anyway.

I think clarity and readability is the most important consideration here.
I'll change if you think the for loop is clearer.